File: packages/sdk-node/src/nium.ts
// ============================================================
// @kangopenbanking/sdk — Nium extended resources
// Aligned with OpenAPI v4.52.0 — /v1/gateway/nium/*
// Additive only (Standing Orders 1, 2, 4).
// ============================================================

import type { KangOpenBanking } from './client';

export type NiumCurrency =
  | 'USD' | 'EUR' | 'GBP' | 'AUD' | 'CAD' | 'SGD' | 'AED' | 'JPY'
  | 'INR' | 'ZAR' | 'HKD' | 'CHF' | 'NZD' | 'SEK' | 'NOK' | 'DKK' | 'CNY';

export type NiumAccountKind = 'virtual' | 'global';

export interface NiumBeneficiary {
  id: string;
  user_id: string;
  beneficiary_name: string;
  account_number: string;
  currency: NiumCurrency | 'XAF';
  bank_name?: string | null;
  bic?: string | null;
  iban?: string | null;
  country?: string | null;
  status: 'active' | 'pending' | 'rejected';
  created_at: string;
}

export interface NiumPayout {
  id: string;
  beneficiary_id: string;
  source_currency: NiumCurrency;
  destination_currency: NiumCurrency | 'XAF';
  source_amount: number;
  destination_amount: number;
  fx_rate: number;
  purpose_code: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  idempotency_key: string;
  created_at: string;
}

export interface NiumConversion {
  id: string;
  from_currency: NiumCurrency;
  to_currency: NiumCurrency;
  from_amount: number;
  to_amount: number;
  fx_rate: number;
  status: 'pending' | 'completed' | 'failed';
  idempotency_key: string;
  created_at: string;
}

export interface NiumRfi {
  id: string;
  subject: string;
  description: string;
  status: 'open' | 'responded' | 'closed';
  requested_at: string;
  responded_at?: string | null;
  documents?: string[];
}

export class BeneficiariesResource {
  constructor(private c: KangOpenBanking) {}
  list(): Promise<{ beneficiaries: NiumBeneficiary[] }> {
    return this.c.request('GET', 'nium-beneficiaries');
  }
  create(params: {
    beneficiary_name: string;
    account_number: string;
    currency: NiumCurrency | 'XAF';
    bic?: string;
    iban?: string;
    bank_name?: string;
    country?: string;
  }): Promise<{ beneficiary: NiumBeneficiary; reused: boolean }> {
    return this.c.request('POST', 'nium-beneficiaries', params);
  }
}

export class PayoutsResource {
  constructor(private c: KangOpenBanking) {}
  list(): Promise<{ payouts: NiumPayout[] }> {
    return this.c.request('GET', 'nium-payouts');
  }
  /** Requires UUID v4 Idempotency-Key. Auto-quotes FX when source ≠ destination. */
  create(params: {
    beneficiary_id: string;
    source_currency: NiumCurrency;
    destination_currency: NiumCurrency | 'XAF';
    source_amount: number;
    purpose_code: string;
    idempotency_key: string;
  }): Promise<{ payout: NiumPayout }> {
    return this.c.request('POST', 'nium-payouts', params, {
      'Idempotency-Key': params.idempotency_key,
    });
  }
}

export class ConversionsResource {
  constructor(private c: KangOpenBanking) {}
  list(): Promise<{ conversions: NiumConversion[] }> {
    return this.c.request('GET', 'nium-conversions');
  }
  create(params: {
    from_currency: NiumCurrency;
    to_currency: NiumCurrency;
    from_amount: number;
    idempotency_key: string;
  }): Promise<{ conversion: NiumConversion }> {
    if (params.from_currency === params.to_currency) {
      throw new Error('from_currency must differ from to_currency');
    }
    return this.c.request('POST', 'nium-conversions', params, {
      'Idempotency-Key': params.idempotency_key,
    });
  }
}

export class RfiResource {
  constructor(private c: KangOpenBanking) {}
  list(status?: 'open' | 'responded' | 'closed'): Promise<{ rfis: NiumRfi[] }> {
    return this.c.request('GET', 'nium-rfi', status ? { status } : undefined);
  }
  respond(params: { rfi_id: string; response: string; document_urls?: string[] }): Promise<{ rfi: NiumRfi }> {
    return this.c.request('POST', 'nium-rfi', params);
  }
}

File: packages/sdk-python/kangopenbanking/nium.py
"""Nium extended resources — Beneficiaries, Payouts, Conversions, RFI.

Aligned with OpenAPI v4.52.0 — /v1/gateway/nium/*.
Additive only (Standing Orders 1, 2, 4).
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

NiumCurrency = Literal[
    "USD", "EUR", "GBP", "AUD", "CAD", "SGD", "AED", "JPY",
    "INR", "ZAR", "HKD", "CHF", "NZD", "SEK", "NOK", "DKK", "CNY",
]
NiumAccountKind = Literal["virtual", "global"]


class BeneficiariesResource:
    def __init__(self, client):
        self._c = client

    def list(self) -> Dict[str, Any]:
        return self._c._request("GET", "nium-beneficiaries")

    def create(
        self,
        *,
        beneficiary_name: str,
        account_number: str,
        currency: str,
        bic: Optional[str] = None,
        iban: Optional[str] = None,
        bank_name: Optional[str] = None,
        country: Optional[str] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {
            "beneficiary_name": beneficiary_name,
            "account_number": account_number,
            "currency": currency,
        }
        for k, v in (("bic", bic), ("iban", iban), ("bank_name", bank_name), ("country", country)):
            if v is not None:
                body[k] = v
        return self._c._request("POST", "nium-beneficiaries", json=body)


class PayoutsResource:
    def __init__(self, client):
        self._c = client

    def list(self) -> Dict[str, Any]:
        return self._c._request("GET", "nium-payouts")

    def create(
        self,
        *,
        beneficiary_id: str,
        source_currency: str,
        destination_currency: str,
        source_amount: float,
        purpose_code: str,
        idempotency_key: str,
    ) -> Dict[str, Any]:
        body = {
            "beneficiary_id": beneficiary_id,
            "source_currency": source_currency,
            "destination_currency": destination_currency,
            "source_amount": source_amount,
            "purpose_code": purpose_code,
            "idempotency_key": idempotency_key,
        }
        return self._c._request(
            "POST", "nium-payouts", json=body,
            headers={"Idempotency-Key": idempotency_key},
        )


class ConversionsResource:
    def __init__(self, client):
        self._c = client

    def list(self) -> Dict[str, Any]:
        return self._c._request("GET", "nium-conversions")

    def create(
        self,
        *,
        from_currency: str,
        to_currency: str,
        from_amount: float,
        idempotency_key: str,
    ) -> Dict[str, Any]:
        if from_currency == to_currency:
            raise ValueError("from_currency must differ from to_currency")
        body = {
            "from_currency": from_currency,
            "to_currency": to_currency,
            "from_amount": from_amount,
            "idempotency_key": idempotency_key,
        }
        return self._c._request(
            "POST", "nium-conversions", json=body,
            headers={"Idempotency-Key": idempotency_key},
        )


class RfiResource:
    def __init__(self, client):
        self._c = client

    def list(self, status: Optional[str] = None) -> Dict[str, Any]:
        params = {"status": status} if status else None
        return self._c._request("GET", "nium-rfi", params=params)

    def respond(
        self, *, rfi_id: str, response: str, document_urls: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"rfi_id": rfi_id, "response": response}
        if document_urls:
            body["document_urls"] = document_urls
        return self._c._request("POST", "nium-rfi", json=body)
